File: Graph/traversal.js
// DFS

// Iterative
let depthFirstPrint = (graph, source) => {
  const stack = [source];
  while(stack.length) {
    let currentNode = stack.pop();
    console.log(currentNode);
    for(let neighbor of graph[currentNode]) {
      stack.push(neighbor);
    }
  }
}

// Recursive
depthFirstPrint = (graph, source) => {
  console.log(source);
  for(let neighbor of graph[source]) {
    depthFirstPrint(graph, neighbor);
  }
}

// BFS
const breadthFirstPrint = (graph, source) => {
  const queue = [source];
  while(queue.length) {
    let currentNode = queue.shift();
    console.log(currentNode);
    for(let neighbor of graph[currentNode]) {
      queue.push(neighbor);
    }
  }
}

const graph = {
  a: ['b', 'c'],
  b: ['d'],
  c: ['e'],
  d: ['f'],
  e: [],
  f: []
}

// Graph and the starting point
console.log('DFS');
depthFirstPrint(graph, 'a');
console.log('BFS');
breadthFirstPrint(graph, 'a')

File: Python/Graph/traversal.py
def depth_first_print(graph, source):
    stack = [source]
    while len(stack) > 0:
        current = stack.pop()
        print(current)
        for neighbor in graph[current]:
            stack.append(neighbor)
            
def depth_first_recursive_print(graph, source):
    print(source)
    for neighbor in graph[source]:
        depth_first_recursive_print(graph, neighbor)

graph = {
  'a': ['b', 'c'],
  'b': ['d'],
  'c': ['e'],
  'd': ['f'],
  'e': [],
  'f': []
}

# Graph and the starting point
print('DFS');
depth_first_print(graph, 'a');
print('DFS Recursive');
depth_first_recursive_print(graph, 'a');
#console.log('BFS');
#breadthFirstPrint(graph, 'a')